src/Observable.ts

```diff
@@ -0,0 +1,492 @@
+import './Symbol';
+
+/**
+ * Handles an individual subscription to an Observable.
+ */
+export interface Subscription {
+	/**
+	 * Whether or not the subscription is closed. Closed subscriptions will not emit values.
+	 */
+	closed: boolean;
+
+	/**
+	 * A function to call to close the subscription. Calling this will call any associated tear down methods.
+	 */
+	unsubscribe: (() => void);
+}
+
+/**
+ * Handles events emitted from the subscription
+ */
+export interface Observer<T> {
+	/**
+	 * Called to handle a single emitted event.
+	 *
+	 * @param {T} value    The value that was emitted.
+	 */
+	next?(value: T): any;
+
+	/**
+	 * An optional method to be called when the subscription starts (before any events are emitted).
+	 * @param observer
+	 */
+	start?(observer: Subscription): void;
+
+	/**
+	 * An optional method to be called if an error occurs during subscription or handling.
+	 *
+	 * @param errorValue    The error
+	 */
+	error?(errorValue: any): any;
+
+	/**
+	 * An optional method to be called when the subscription is completed (unless an error occurred and the error method was specified)
+	 *
+	 * @param completeValue    The value passed to the completion method.
+	 */
+	complete?(completeValue?: any): void;
+}
+
+/**
+ * An object used to control a single subscription and an observer.
+ */
+export interface SubscriptionObserver<T> {
+	/**
+	 * Whether or not the subscription is closed.
+	 */
+	readonly closed: boolean;
+
+	/**
+	 * Emit an event to the observer.
+	 *
+	 * @param value    The value to be emitted.
+	 */
+	next(value: T): any;
+
+	/**
+	 * Report an error. The subscription will be closed after an error has occurred.
+	 *
+	 * @param errorValue    The error to be reported.
+	 */
+	error(errorValue: any): any;
+
+	/**
+	 * Report completion of the subscription. The subscription will be closed, and no new values will be emitted,
+	 * after completion.
+	 *
+	 * @param completeValue?    A value to pass to the completion handler.
+	 */
+	complete(completeValue?: any): void;
+}
+
+export interface Subscriber<T> {
+	(observer: SubscriptionObserver<T>): (() => void) | void | { unsubscribe: () => void };
+}
+
+/**
+ * An object that implements a Symbol.observerable method.
+ */
+export interface ObservableObject {
+	[Symbol.observable]: () => any;
+}
+
+namespace Shim {
+	/*
+	 * Decorator to mark a single method/property as non-enumerable. ES spec requires pretty much every
+	 * method or property in Subscription, Observable, and SubscriptionObserver to be non-enumerable.
+	 */
+	function nonEnumerable(target: any, key: string | symbol, descriptor: PropertyDescriptor) {
+		descriptor.enumerable = false;
+	}
+
+	/*
+	 * Create a subscription observer for a given observer, and return the subscription.  The "logic" for Observerables
+	 * is in here!
+	 */
+	function startSubscription<T>(executor: Subscriber<T>, observer: Observer<T>): Subscription {
+		let closed = false;
+		let cleanUp: any;
```

Comment:
So `cleanUp` is either `undefined` or a method with no parameters and no return value, so I've typed it as such (not just the return value of `executor`).